kotlin-native/backend.native/compiler/ir/backend.native/src/org/jetbrains/kotlin/backend/konan/objcexport/ObjCExportNamer.kt

```diff
@@ -902,6 +929,84 @@ private fun ObjCExportMapper.canHaveSameName(first: PropertyDescriptor, second:
     return bridgePropertyType(first) == bridgePropertyType(second)
 }
 
+private class ObjCName(
+        private val kotlinName: String,
+        private val objCName: String?,
+        private val swiftName: String?,
+        val isExact: Boolean
+) {
+    // TODO: Prevent mangling when objCName or swiftName is provided
+
+    fun asString(forSwift: Boolean): String = swiftName.takeIf { forSwift } ?: objCName ?: kotlinName
+
+    fun asIdentifier(forSwift: Boolean, default: () -> String = kotlinName::toIdentifier): String =
+            swiftName.takeIf { forSwift } ?: objCName ?: default()
+}
+
+private fun DeclarationDescriptor.getObjCName(): ObjCName {
+    var objCName: String? = null
+    var swiftName: String? = null
+    var isExact = false
+    annotations.findAnnotation(KonanFqNames.objCName)?.let { annotation ->
+        objCName = annotation.argumentValue("name")?.value?.cast()
+        swiftName = annotation.argumentValue("swiftName")?.value?.cast()
+        isExact = annotation.argumentValue("exact")?.value?.cast() ?: false
+    }
+    return ObjCName(name.asString(), objCName, swiftName, isExact)
+}
+
+private fun CallableDescriptor.getObjCName(): ObjCName =
+        overriddenDescriptors.firstOrNull()?.getObjCName() ?: (this as DeclarationDescriptor).getObjCName()
+
+private fun ParameterDescriptor.getObjCName(): ObjCName {
+    val callableDescriptor = containingDeclaration as? CallableDescriptor ?: return (this as CallableDescriptor).getObjCName()
+    fun CallableDescriptor.getBase(): CallableDescriptor = overriddenDescriptors.firstOrNull()?.getBase() ?: this
+    val baseCallableDescriptor = callableDescriptor.getBase()
+    if (callableDescriptor.extensionReceiverParameter == this) {
+        return (baseCallableDescriptor.extensionReceiverParameter as CallableDescriptor).getObjCName()
+    }
+    val parameterIndex = callableDescriptor.valueParameters.indexOf(this)
+    if (parameterIndex != -1) {
+        return (baseCallableDescriptor.valueParameters[parameterIndex] as CallableDescriptor).getObjCName()
+    }
+    error("Unsupported ObjCName annotated parameter $this")
```

Comment:
The error message is a bit misleading -- it might happen even if no parameters are annotated with `@ObjCName`, because this code is used in any case.